S3TransferUtilitySample/src/com/amazonaws/demo/s3transferutility/UploadActivity.java

```diff
@@ -519,85 +517,68 @@ private void beginUploadInBackground(String filePath) {
-                final String[] split = docId.split(":");
-                final String type = split[0];
-                if ("image".equals(type)) {
-                    uri = MediaStore.Images.Media.EXTERNAL_CONTENT_URI;
-                } else if ("video".equals(type)) {
-                    uri = MediaStore.Video.Media.EXTERNAL_CONTENT_URI;
-                } else if ("audio".equals(type)) {
-                    uri = MediaStore.Audio.Media.EXTERNAL_CONTENT_URI;
+    private File getFile(Uri uri) throws FileNotFoundException {
+        final InputStream in = getContentResolver().openInputStream(uri);
+        OutputStream out = null;
+        try {
+            final File file = new File(getCacheDir(), getFileName(uri));
+            out = new FileOutputStream(file, false);
+            byte[] buffer = new byte[1024];
+            for (int len; (len = in.read(buffer)) != -1; ) {
+                out.write(buffer, 0, len);
+            }
+            return file;
+        } catch (IOException e) {
+            e.printStackTrace();
+        } finally {
+            try {
+                if (in != null) {
+                    in.close();
                 }
-                selection = "_id=?";
-                selectionArgs = new String[] {
-                        split[1]
-                };
+                if(out != null) {
+                    out.close();
+                }
+            } catch (IOException e) {
+                e.printStackTrace();
             }
         }
+        return null;
+    }
+
+    private String getFileName(Uri uri) {
         if ("content".equalsIgnoreCase(uri.getScheme())) {
-            String[] projection = {
-                    MediaStore.Images.Media.DATA
+            final String[] projection = {
+                    MediaStore.Images.Media.DISPLAY_NAME
             };
-            Cursor cursor = null;
+            final String docId = DocumentsContract.getDocumentId(uri);
+            String selection = null;
+            String[] selectionArgs = null;
+            if(docId.contains(":")) {
+                final String[] split = docId.split(":");
+                selection = "_id=?";
+                selectionArgs = new String[]{
+                        split[1]
+                };
+            }
             try {
-                cursor = getContentResolver()
-                        .query(uri, projection, selection, selectionArgs, null);
-                int column_index = cursor.getColumnIndexOrThrow(MediaStore.Images.Media.DATA);
+                Cursor cursor = getContentResolver().query(uri, projection, selection, selectionArgs, null);
+                int column_index = cursor.getColumnIndexOrThrow(MediaStore.Images.Media.DISPLAY_NAME);
                 if (cursor.moveToFirst()) {
                     return cursor.getString(column_index);
                 }
             } catch (Exception e) {
             }
-        } else if ("file".equalsIgnoreCase(uri.getScheme())) {
```

Comment:
is this the remnant of our "if SDK<19" bits of code in the app?

Reply:
we should add a new chore to kind of "clean up" our code. I noticed a lot of random bits of code in the sample app that don't seem to be doing anything like:
```
if (API >= 19) {
  // do stuff
} else {
  // do stuff... or not because minSdk = 23
}
```

Reply:
Hmm, I just noticed minSdkVersion = 23, and the following comment explaining why:
```
 // aws-android-sdk-mobile-client depends on aws-android-sdk-auth-ui which requires a minSdkVersion of 23  
```

Are there any plans to ever lower this to 16 (min requirement from our [docs](https://docs.amplify.aws/lib/storage/getting-started?platform=android))?  

I ask because using `try-with-resources` as @jamesonwilliams suggested requires API 19 and above.

Reply:
@richardmcclellan  That _used to be_ true. Now:
> In addition to the Java 8 language features and APIs above, Android Studio 3.0 and later extends support for try-with-resources to all Android API levels.

Reply:
Got it!  I realized I need to add this to the build.gradle in order for it to work with a lower minSdkVersion:
```
compileOptions {
        sourceCompatibility JavaVersion.VERSION_1_8
        targetCompatibility JavaVersion.VERSION_1_8
    }
```